presto-main/src/main/java/com/facebook/presto/execution/scheduler/group/DynamicLifespanScheduler.java

```diff
@@ -107,39 +132,62 @@ public void onLifespanExecutionFinished(Iterable<Lifespan> newlyCompletelyExecut
         synchronized (this) {
             for (Lifespan newlyCompletelyExecutedDriverGroup : newlyCompletelyExecutedDriverGroups) {
                 checkArgument(!newlyCompletelyExecutedDriverGroup.isTaskWide());
-                recentlyCompletelyExecutedDriverGroups.add(newlyCompletelyExecutedDriverGroup);
+                availableTasks.enqueue(taskByDriverGroup[newlyCompletelyExecutedDriverGroup.getId()]);
                 totalLifespanExecutionFinished++;
+                int driverGroupId = newlyCompletelyExecutedDriverGroup.getId();
+                runningDriverGroupIdsByTask[taskByDriverGroup[driverGroupId]].remove(driverGroupId);
             }
             newDriverGroupReady = this.newDriverGroupReady;
         }
         newDriverGroupReady.set(null);
     }
 
+    @Override
+    public boolean retryFailedTask(int taskId, List<SourceScheduler> sourceSchedulers)
+    {
+        checkState(initialScheduled);
+
+        synchronized (this) {
+            this.failedTasks.add(taskId);
+            if (failedTasks.size() > nodeByTaskId.size() * maxFailedTaskPercentage) {
+                return false;
+            }
+            for (int driverGroupId : runningDriverGroupIdsByTask[taskId]) {
+                sourceSchedulers.forEach(sourceScheduler -> sourceScheduler.rewindLifespan(Lifespan.driverGroup(driverGroupId), partitionHandles.get(driverGroupId)));
```

Comment:
I would just use plain for-loop here. `foreach` doesn't seem to make anything easy to understand :)